docarray/array/storage/base/getsetdel.py

```diff
@@ -173,15 +186,26 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
         for _id, doc in zip(ids, docs):
             self._set_doc_by_id(_id, doc)
 
-    def _update_subindices_set(self, ids, docs):
-        if getattr(self, '_subindices', None):
-            for selector, da in self._subindices.items():
-                old_ids = DocumentArray(self[ids])[
-                    selector, 'id'
-                ]  # hack to get the Document['@c'] without having to do Document.chunks
-                with da:
-                    del da[old_ids]
-                    da.extend(DocumentArray(docs)[selector])  # same hack here
+    def _update_subindices_set(self, set_index, docs):
+        subindices = getattr(self, '_subindices', None)
+        if not subindices:
+            return
+        if isinstance(set_index, tuple):  # handled later in recursive call
+            return
+        if isinstance(set_index, str) and set_index.startswith('@'):
+            # 'nested' (non root-level) set, update entire subindex directly
+            _check_valid_values_nested_set(DocumentArray(self[set_index]), docs)
```

Comment:
do we need to wrap into `DocumentArray` ? it's cast to DocumentArray inside the check function

Reply:
You are right, I'll change it